### PR TITLE
-SIPC-/time: some bug fixes and new blocks

### PR DESCRIPTION
**Pull request by SharkPool-SP**
Moved From: https://github.com/TurboWarp/extensions/pull/752#issuecomment-1689271015
(OG .js base file was outdated)

Adds several new calculation blocks for time!
Note: Fixed Several Bugs from Before AND added new blocks!!...
Screenshot 2023-09-16 212935

### CODE REVIEW

File: extensions/-SIPC-/time.js [hunk 1]
```diff
@@ -70,20 +75,124 @@
...
+                type: Scratch.ArgumentType.STRING,
+                menu: "Time",
+                defaultValue: "day",
+              },
+            },
+          },
+          {
+            opcode: "calculatetimedurationfromtime",
+            blockType: Scratch.BlockType.REPORTER,
+            text: "time duration from [START_HOUR]:[START_MINUTE] to current time in [TIME_MENU]",
```
Comment (GarboMuffin) [inline extensions/-SIPC-/time.js:101]: you dont need two inputs for the time here

Comment (GarboMuffin) [reply extensions/-SIPC-/time.js:101]: you should also rename this and the above to be more like the below

"difference between 12:30 and current time in days"

File: extensions/-SIPC-/time.js [hunk 2]
```diff
@@ -138,6 +247,94 @@
...
+      const days = Math.floor(hours / 24);
+
+      const timeString = `${hours % 24}:${minutes % 60}:${seconds % 60}`;
+      return timeString;
+    }
+
+    daysinmonth(args) {
+      const month = args.MONTH;
+      const year = args.YEAR ? args.YEAR : null;
+      const date = new Date(year, this._getMonthIndex(month) + 1, 0);
```
Comment (GarboMuffin) [inline extensions/-SIPC-/time.js:317]: add better error handling if the month isn't recognized. better to return something that is obviously wrong than something that is subtly incorrect

File: extensions/-SIPC-/time.js [hunk 3]
```diff
@@ -138,6 +247,94 @@
...
+        case "month":
+          return Math.floor(timeDiff / (1000 * 60 * 60 * 24 * 30));
+        case "day":
+          return Math.floor(timeDiff / (1000 * 60 * 60 * 24));
+        case "hour":
+          return Math.floor(timeDiff / (1000 * 60 * 60));
+        case "minute":
+          return Math.floor(timeDiff / (1000 * 60));
+        case "second":
+          return Math.floor(timeDiff / 1000);
```
Comment (GarboMuffin) [inline extensions/-SIPC-/time.js:266]: must always return *some* value

**Comment by SharkPool-SP** [comment]
Screenshot 2023-09-06 222521
Added 2 new Blocks!!!

**Comment by GarboMuffin** [comment]
you're probably gonna be pretty unhappy:

 - removed countdown blocks: they are redundant, one can just invert the difference block that already exists
 - removed blocks that have unintuitive dependence on the unix epoch: the expression "January 1, 1970 + January 27, 2024 = January 27, 2024" just doesn't really make sense. If someone wants to unix them as unix timestamps, the blocks are already there for them.
 - Similarly "convert January 27, 2024 to years = 54.11020848940893" just doesn't seem very useful, and again its already easy for someone to compute with the existing blocks

**Comment by SharkPool-SP** [comment]
> you're probably gonna be pretty unhappy:
> 
> 
> 
>  - removed countdown blocks: they are redundant, one can just invert the difference block that already exists
> 
>  - removed blocks that have unintuitive dependence on the unix epoch: the statement that "January 1, 1970 + January 27, 2024 = January 27, 2024" just doesn't really make sense. If someone wants to unix them as unix timestamps, the tools are already there for them to do that themselves.
> 
>  - Similarly the statement "convert January 27, 2024 to years = 54.11020848940893" just doesn't seem very useful, and again its already easy for someone to compute with the existing blocks

Fine by me

**Pull request merged**